src/agent/src/device.rs

```diff
@@ -671,31 +677,27 @@ async fn vfio_device_handler(
         if vfio_in_guest {
             pci_driver_override(SYSFS_BUS_PCI_PATH, guestdev, "vfio-pci")?;
 
-            let devgroup = pci_iommu_group(SYSFS_BUS_PCI_PATH, guestdev)?;
-            if devgroup.is_none() {
-                // Devices must have an IOMMU group to be usable via VFIO
-                return Err(anyhow!("{} has no IOMMU group", guestdev));
-            }
+            let devgroup = pci_iommu_group(SYSFS_BUS_PCI_PATH, guestdev)?
+                .ok_or_else(|| anyhow!("{} has no IOMMU group", guestdev))?;
 
-            if group.is_some() && group != devgroup {
+            if let Some(g) = group {
                 // If PCI devices associated with the same VFIO device
                 // (and therefore group) in the host don't end up in
                 // the same group in the guest, something has gone
-                // horribly wrong
-                return Err(anyhow!(
-                    "{} is not in guest IOMMU group {}",
-                    guestdev,
-                    group.unwrap()
-                ));
+                // horribly wrong.
+                if g != devgroup {
+                    return Err(anyhow!("{} is not in guest IOMMU group {}", guestdev, g));
+                }
             }
 
-            group = devgroup;
+            group = Some(devgroup);
         }
     }
 
     if vfio_in_guest {
         // If there are any devices at all, logic above ensures that group is not None
-        let group = group.unwrap();
+        let group = group.ok_or_else(|| anyhow!("failed to get VFIO group: {:?}"))?;
+
```

Comment:
Again, we've ensured this is the case in the code enough, so reporting an error is misleading.  A failure here is definitely a code bug, hence an assert (which unwrap() is a specialized form of) is appropriate.